internal/server/recon/recognize.go

```diff
@@ -73,6 +73,138 @@ func RecognizePlaces(
+		}
+		sort.Slice(spans, func(i, j int) bool {
+			return len(spans[i]) > len(spans[j])
+		})
+
+		// iterate through the resolved spans and add the resolved items to the
+		// response
+		queryRespItems := []*pb.RecognizePlacesResponse_Item{{Span: query}}
+		for _, span := range spans {
+			// iterate through the response items to see where this span came from
+			for i, item := range queryRespItems {
+				// if there were Places already detected in this response item, move on
+				// to the next item
+				if len(item.Places) > 0 {
+					continue
+				}
+				// try to split the span of the response item into a list of spans
+				// containing the current span as its own span. if not possible, move on
+				// to the next item.
+				newSpans := getNewSpanList(item.Span, span)
+				if len(newSpans) == 0 {
+					continue
+				}
+				// replace the current response item with a new list of items containing
+				// the new resolved entity
+				newItems := []*pb.RecognizePlacesResponse_Item{}
+				for _, newSpan := range newSpans {
+					trimmedSpan := strings.TrimSpace(newSpan)
+					if len(trimmedSpan) < 1 {
+						continue
+					}
+					if trimmedSpan == span {
+						newItems = append(newItems, span2item[span])
+					} else {
+						newItems = append(newItems, &pb.RecognizePlacesResponse_Item{Span: trimmedSpan})
+					}
+				}
+				newItems = append(newItems, queryRespItems[i+1:]...)
+				queryRespItems = append(queryRespItems[:i], newItems...)
+			}
+		}
+		resp.QueryItems[query] = &pb.RecognizePlacesResponse_Items{Items: queryRespItems}
+	}
+	return resp, nil
+}
+
+func getNewSpanList(origSpan string, splitSpan string) []string {
+	newSpans := strings.Split(origSpan, splitSpan)
+	// if split span is not found in original span, return empty list
+	if len(newSpans) < 2 {
+		return []string{}
+	}
+	// if the item before splitSpan is not empty and does not end in a space,
+	// assume a word got split in the middle and return empty list.
+	if len(newSpans[0]) > 0 && !strings.HasSuffix(newSpans[0], " ") {
+		return []string{}
+	}
+	// if the item after splitSpan is not empty and does not start with a space,
+	// assume a word got split in the middle and return empty list.
+	if len(newSpans[1]) > 0 && !strings.HasPrefix(newSpans[1], " ") {
+		return []string{}
+	}
+	return []string{strings.TrimSpace(newSpans[0]), splitSpan, strings.TrimSpace(newSpans[1])}
+}
+
+// Takes a query and returns a map of id to use for resolution to the
+// original span (part of the query) for that id.
+func getId2Span(query string) map[string]map[string]struct{} {
+	id2spans := map[string]map[string]struct{}{}
+	spanTokens := strings.Split(query, " ")
+	for i := range spanTokens {
+		span := ""
+		// make n-grams from the span tokens
+		for j := i; j < len(spanTokens); j++ {
+			span = span + " " + spanTokens[j]
+			span = strings.TrimSpace(span)
+			id := strings.ReplaceAll(span, " ,", "")
+			id = strings.ReplaceAll(id, ",", "")
+			id = strings.ReplaceAll(id, "^", "")
```

Comment:
Consider having a helper for "name to reconName", and commenting that this should following the corresponding logic in the flume pipeline (link to the func-name, g3 url)?

Reply:
done